Refactoring and improvements

piegamesde opened the pull request:
See commit log. Generally speaking, this changes nothing on the protocol itself. Just the documentation.

piegamesde committed:
Server protocol: Improve the `welcome` message

- Deprecate the `current_cli_version` key
- Improved wording, added type information

piegamesde committed:
Remove Sphinx related files

We want this repository to be agnostic of any site generators or similar.
Information related to rendering the markdown as web page should be stored
externally.

piegamesde committed:
Delete new-protocol.svg

It doesn't seem to be reference anywhere and also looks like a prototype/sketch.

piegamesde committed:
Remove Python-specific documentation

Closes #30.

Those files also contained a couple of paragraphs – mainly the general introduction –
that will have to be rewritten in a client-agnostic style.

piegamesde committed:
Rename attacks.md → security.md

piegamesde committed:
Add new landing page

piegamesde committed:
Update the security discussion

Rewrote the existing topics. Tried to minimize the discussion of any hypotheticals
and future work, as this should be done outside of the document itself.

Added privacy sections discussing file size metadata and IP addresses.

meejah commented on security.md:
```diff
@@ -0,0 +1,94 @@
+be characteristic for its content. (Imagine people using Magic Wormhole to send
+some specific movie.)
+
+Thanks to the separation between rendezvous server and relay server, forcing the
```
Other places I believe we'd agreed that "mailbox" and "transit" are less-confusing names; perhaps we should use those here now too.

meejah commented on security.md:
```diff
@@ -0,0 +1,94 @@
+
+As usual in cryptography, communication metadata cannot easily be protected. Any
+involved server will know the timestamp and size of each message it relays. This
+usually is not much of an issue, except for file transfer where the file size may
```
Well, _really_ depends on perspective? (What's a case where it isn't an issue?)

piegamesde replied:
Most of the time actually, the file size does not contain that much information: an attacker gains nothing of value from the fact that my configuration file has 45219 bytes (ignoring for now that the protocol may round numbers up to some block size for encryption).

meejah replied:
File-sizes are actually extremely unique (unfortunately). This e.g. forms part of the basis of "fingerprinting" Web pages, which is extremely successful (e.g. allowing one to answer the question "what website is this TLS connection visiting" even on networks like Tor).

So, I would assume this can be extended to pattern-match on message flows in wormhole as well (allowing fingerprinting of any protocol). I guess it's the "not much of an issue" comment that I'm commenting on here...

meejah replied:
Trying to be more-clear: leaking metadata is bad.

So, it's still bad for almost any "application protocol" I can think of, and it should remain clear that Wormhole does NOT protect against that -- allowing passive observers (including the mailbox server) to learn many things about your protocol even though the *contents* remain encrypted an unview-able by the server.

meejah commented on security.md:
```diff
@@ -0,0 +1,94 @@
+However, keep in mind that it is easy to host a relay server, so it might still
+leak your public IP address.
+
+If you do not trust the servers with knowing your IP address either, use
```
We generally try to avoid this word .. better is "expose"; that is something like "if you do not want to expose your IP to ..."

piegamesde replied:
Which word?

meejah replied:
Ah, sorry ;) "trust"

piegamesde replied:
Oh that's interesting, why? Because I see trust as an essential part of threat modeling and assessing individual security risks.

meejah replied:
I think this was via Tor people I'm thinking of? You aren't actually "trusting" the server, so the "expose" I guess is trying to get at what's actually happening. That is, that you reveal your IP address to them -- I don't think "trust" enters into it per se. Maybe "reveal" is a good one too?

I believe the logic is that it doesn't matter if you "trust" the server, or the people running the server -- the important thing is what information you reveal. For example, even if they are trustworthy they can themselves be compromised or forced to reveal any information they have (subpoena, etc)

meejah commented on security.md:
```diff
@@ -0,0 +1,94 @@
+If you do not trust the servers with knowing your IP address either, use
+[Tor](https://torproject.org/) to use Magic Wormhole anonymously. The Transit
+protocol has Tor support built-in, which then makes use of the Tor network itself
+as the relay.
```
I'm not clear what you mean here? The transit is still involved, because we lack support for Onion services (this part is mentioned in a file you've deleted -- perhaps just including that stuff here would be better).

piegamesde replied:
I was referring to the Tor part which should be documented in the Transit protocol specification. I'll make it more explicit

meejah replied:
It's still using the transit server though, so the part about "use the Tor network itself as the relay" isn't really correct (or I don't understand what you mean here).

piegamesde replied:
Yes you can always connect to a relay from behind Tor, but that's not what I'm talking about. The transit protocol specifies "Tor hints" which point to onion domains, of which the idea is to spin up a hidden service and send that over as a hint. In that case, there is a "direct" connection with respect to the Tor network. But from an IP perspective, the Tor network merely acts as a relay.

meejah replied:
Oh, it actually spins up Onion services?? Cool .. I guess it's been a while since I looked in there very deeply.

I suppose you could consider that Tor is basically acting "as" the relay in that case -- although I guess you could always say Tor is "a relay" in that sense. You could also view those as "merely" direct connections (albeit via Tor) since the semantics of Onion domains give you assurance you're connecting to the right computer.

Okay, I guess it's fine as-is.

It might still be worth distinguishing between "use Tor, but still via the Transit" vs. "use only Tor"...? (The former thing is possible, right?)

piegamesde replied:
> Oh, it actually spins up Onion services?

Actually I don't know ^^ I know for sure that the protocol supports this and that this definitely was the original plan. But to which extent it has been implemented in the Python client, no idea. (IMO the latter question does not matter that much because this document should mostly speak from a protocol perspective.)

> It might still be worth distinguishing between "use Tor, but still via the Transit" vs. "use only Tor"...? (The former thing is possible, right?)

Yes, I'll try to work that in

meejah reviewed (changes requested):
A lot of this seems straightforward, but I think "security" doesn't accurately re-create the scenarios described in "attacks", and I think we lose a little by deleting the "approaches considered but not tried" (see inline comments).

piegamesde commented on security.md:
```diff
@@ -0,0 +1,94 @@
+receiver's attempt will fail, such an attack will be noticeable.
+
+If the attacker *additionally* has the capability of intercepting and changing
+network traffic ("MitM"), they might be able to fake arbitrary interactions
```
Oh, I see. I interpreted the wording wrong then, will update

piegamesde commented:
Thanks for the review. Ideally, none of the important information should be lost in the rewrite, so I'll go and add the missing bits

piegamesde commented on security.md:
```diff
@@ -0,0 +1,94 @@
+Both cases can be mitigated by enabling the verifier feature, in which case
+even successful attacks can be detected.
+
+In general, this is the same threat model as with ordinary key exchanges, except
```
Replacing this with a Diffie-Hellman comparison, as it turns out there are more different kinds of key exchanges than I'd remembered.

piegamesde commented on security.md:
```diff
@@ -0,0 +1,94 @@
+enumerate all nameplates to disrupt them, but even without it the name space is
+sufficiently small (by design, as we want short codes) to brute force.
+
+There are several ideas to defend against this, however the only one already
```
Are you talking only about rejected ideas, or also about ideas under consideration/planning that we might add in the future?

meejah replied:
After more thought it's fine to just leave them out; we can add more if / when there's more-concrete plans etc.

meejah commented on security.md:
```diff
@@ -3,23 +3,23 @@
+machine-in-the-middle (MitM) attack. For this, they quickly reconnect with
+the same code in order to connect with the second peer. With that, they get a
+secure channel to each side while being able to read and modify the connection's
+content.
```
:+1:

piegamesde commented:
I added a commit that deprecates the `current_cli_version` field in the `welcome` message from the server. Of course the server will continue to provide it anyways, but it has no value to any other client other than the Python implementation. Should it be desirable we can create a new replacement for the other clients in the future.

meejah commented:
I don't actually see anywhere in the current mailbox code where it sends a `current_cli_version` (all I did is searched for that string but don't find anything). Anyway, fine to mention it here as "deprecated" (as you point out it's not going to work as-is for multiple kinds of clients anyway).

meejah commented on index.md:
```diff
@@ -0,0 +1,28 @@
+to Magic Wormhole, which are not specific to a single client or implementation.
+It assumes the reader is already familiar with the general Magic Wormhole concept.
+
+The most important component is the **rendezvous server**. There are two aspects
```
maybe add "aka mailbox server" to avoid confusion (while we're still .. using both?)

meejah commented on index.md:
```diff
@@ -0,0 +1,28 @@
+secure way of exchanging messages.
+
+Using the established low-bandwidth secure channel, both sides then negotiate a
+ecure high-bandwidth channel, called [**transit**](./transit.md). The transit
```
typo -> secure

meejah commented on index.md:
```diff
@@ -0,0 +1,28 @@
+Using the established low-bandwidth secure channel, both sides then negotiate a
+ecure high-bandwidth channel, called [**transit**](./transit.md). The transit
+protocol describes how both sides establish a direct connection, how a **relay
+server** may be used as fallback, and the cryptography used to make that
```
maybe add "aka Transit server" (similar reasoning to above)

meejah commented on index.md:
```diff
@@ -0,0 +1,28 @@
+makes it possible for applications to replace the traditional code exchange
+with sharing a link or QR code.
+
+Security thread models and privacy considerations are discussed in
```
typo: thread -> threat

meejah commented on security.md:
```diff
@@ -0,0 +1,97 @@
+Compared to password strengths we are used to nowadays, 16 bits may seem scarily
+small, even knowing that an attacker only has one try to guess the password. But
+to put it into perspective: As a user, you'd have to experience 32000 failed
+transfers because of an attack in order to having a 50% chance of a file being
```
typo/grammar: "having" -> "have"

meejah commented on security.md:
```diff
@@ -0,0 +1,97 @@
+intercepted. From an attacker's perspective, guessing thousands of codes is not
+difficult, but there is no possibility of targetting any individual connection,
+making any gained data fairly useless. From a service perspective, any successful
+attack implies 65536 failed attempts on average, which will quickly be noticed
```
I think you mean ~32000 here (to match above)...?

piegamesde replied:
Nope, should be correct. The one is the total average, the other one is only up to 50% probability. Something something Birthday paradox
(The probability of having a cracked password after 65536 attempts is less than one, but the probability of needing less than that also exists so it exactly cancels out on average)

meejah replied:
Can you explain further? I don't immediately see how the Birthday Paradox is relevant here?

piegamesde replied:
Okay thought about this more and it's not the Birthday paradox. So IMO 1/65536 still is correct, but one needs a different calculation for the other number.

I end up at (1 − 1/65536)^45000 = 0.503258791 so if the calculation is correct one would need 45000 guesses for a 50% success probability.

Fxrh replied:
After 32768 attempts, you have an *expected* number of successful guesses that is ``0.5``. However, this number also includes events where you guess the right key multiple times, which doesn't really help you as an attacker. Therefore, the number of attempts you need to have a 50% chance of getting *at least* one correct guess is higher (and seems to be around ``45000``, as @piegamesde showed).

piegamesde replied:
Addendum: this still works even in our model where the attacker only gets a guess at each code. In that case, I think it becomes "the probability of (not) being hacked after 45000 failed exchanges is 0.5"

So in the end I'd say both numbers are correct, they just answer different questions.

meejah replied:
@Fxrh that part I don't understand: there's only one guess available per setup attempt, so what does "guess the right key multiple times" mean here..?

meejah replied:
I think the text could still benefit from more words, then? (And 65536 isn't either of these numbers, so that should change, yes?)

meejah commented on security.md:
```diff
@@ -0,0 +1,97 @@
+See the server protocol specification for details.
+
+Currently the best defense against this is to use high-entropy nameplates which
+cannot easily be enumerated at the cost of longer codes. This has not be
```
If we're going to leave out speculation, maybe this too? (I'd have to do a lot more thinking to convince myself this is "the best" for example). Also typo: "This has not be" -> "This has not been"

piegamesde replied:
Yes, we can leave it out for now. But actually I have some old proposal for "high entropy nameplates" lying around, I could dig it out again and do this, should not be too hard in general.

meejah replied:
If users can exchange high-entropy nameplates, can't they just exchange a key directly?
(I guess I'm assuming "high entropy" here means "32 bytes" ...)

meejah commented on security.md:
```diff
@@ -0,0 +1,97 @@
+However, keep in mind that it is easy to host a relay server, so it might still
+leak your public IP address.
+
+If you do not want to expose your IP address to any of the servers, use
```
Maybe "any of the computers"? (To me in context "server" here means mailbox or transit .. but we want to make clear that you can avoid exposing your IP to the peer too).

meejah replied:
(At least on the Python side I'm pretty sure we _don't_ enumerate local IPs etc when Tor mode is turned on...? I'd have to double-check that, but that's what it should do)

meejah commented on server-protocol.md:
```diff
@@ -72,19 +72,18 @@ messages. Clients must ignore unrecognized message types from the Server.
-might influence its operation. The Python client currently reacts to the
-following keys (and ignores all others):
+might influence its operation. Clients should look out for the following fields,
+and handle them accordingly, if present:
```
Maybe add "clients should simply ignore other fields they don't understand"...?

piegamesde replied:
I thought this was kind of implicit somehow, but no harm done in stating it explicitly.

meejah replied:
:+1: yeah, "explicit usually better" IME.

meejah reviewed (approved):
Overall looks good .. some inline typo/other fixups.

All the new words include "rendezvous" and "relay" which I thought we agreed to avoid in favour of "mailbox" and "transit" (or "transit relay") respectively..? At the very least we should mention the other terms to avoid confusion I feel like.

felinira reviewed (approved):
Thanks for cleaning this up

piegamesde committed:
Server protocol: Rendezvous → Mailbox

Slowly make the naming more consistent.

piegamesde commented:
Unless anybody has strong objections, I'm inclined to merge this now. As this is about documentation and does not make any substantial changes to the protocol, I don't think this requires doing an FCP and stuff.

meejah commented:
I haven't re-reviewed since the force-push, and I think at least the 65536 number should be changed right? (to 32k or 45k depending on how you word it?)

piegamesde commented:
The force-push only added " at least once" as a clarification, see https://github.com/magic-wormhole/magic-wormhole-protocols/compare/95a1e65b510c88ebf552cd057ed27de26011456e..1286cb4159a048918bd2cfdff54708162da7d0ca.

The 65536 should be correct, because it is an average/expected value. (Okay actually it should be 65535, but that's not your point.) After 32768 attempts you'd expect 0.5 successes *on average*, which is not the same as a 50% probability for a single success.

meejah commented:
> The force-push only added " at least once" as a clarification, see https://github.com/magic-wormhole/magic-wormhole-protocols/compare/95a1e65b510c88ebf552cd057ed27de26011456e..1286cb4159a048918bd2cfdff54708162da7d0ca.

It would be a _lot_ easier to see this without force-pushes ...

> The 65536 should be correct, because it is an average/expected value. (Okay actually it should be 65535, but that's not your point.) After 32768 attempts you'd expect 0.5 successes _on average_, which is not the same as a 50% probability for a single success.

I'm more confused then, because the text says "any successful attack implies 65536 failed attempts on average, which will quickly be noticed" but the discussion justifies 32k or 45k. Surely on average there are less failures than that? (Or, please explain in the text).

piegamesde commented:
How about "From a perspective of the service, it will see on average one success and 65535 failures for every 2^16 attempts. Therefore, brute force attacks are inherently very disruptive of the service, and thus easy to detect."

> Surely on average there are less failures than that?

No, the chance of guessing a password with 16 bits strength in a single attempt is precisely 2^-16.

wuan commented:
We could simply say that an attacker would need to do more than 300k attempts to be at least 99% sure to have one successful attempt.

2^16 attempts would only lead to 63% probability, which is not really high.

maralorn commented:
> We could simply say that an attacker would need to do more than 300k attempts to be at least 99% sure to have one successful attempt (Or more than 196k for at least 95%).

I wouldn’t recommend that. This looks like we are putting the bar too high. Like, “Yeah, they might hack the connection. But they can’t be sure to definitely hack the connection, so we are fine …”

wuan commented:
> I wouldn’t recommend that. This looks like we are putting the bar too high. Like, “Yeah, they might hack the connection. But they can’t be sure to definitely hack the connection, so we are fine …”

An attacker who wants to have an acceptable probabilty (> 95%) to succeed will need to do that many attempts.

meejah commented:
> There is an extremely small probability to guess the password already on the initial attempt. This could be overcome by adding a verification message from the receiver back to the sender which can be used to identify the receiver before the data is transmitted.

This feature already exists.

meejah commented:
The chance of rolling a "1" on a d6 is 1/6 .. but we don't say that "in order to roll a 3 there will be on average 5 failures". (Or at least, this framing isn't familiar to me  ... but statistics classes were a *long* time ago ;).

Put another way: the attacker will not keep guessing once they have succeeded, so *on average* they will only need to try half that number of times. Only the least-lucky attacker will have 65535 failures first (well, a strictly _least_ lucky attacker will have an infinite number of failures if we channel H2G2 I suppose because they're so unlucky they will never succeed ;) ).

Anyway, your last suggestion sounds more clear (at least it highlights that the service should see a lot of failures if someone is doing a lot of guessing). I still feel like it's giving an incorrect impression that an attacker will *definitely* steal one connection after 65k tries (or that they need to on average try 65k times to get one, which they don't). That is, by your own reasoning from the "failure" side it should be 45425 (i.e. the first time the chance of failure goes below 50%) so why are you now using 65k here? If you have different reasoning, please respond in the thread about that.

(But, if you wanted to look from the failure way, after 65536 guesses there's still a decent chance of not getting any correct: if we insist on the attacker being unsuccessful every time, they have a (65535/65536) ^ 65536 = 0.37 chance of continuing to be unsuccessful after 65k guesses).

No matter what, I think the text should include some math at least as a footnote to justify whatever number is put here, and why. The important takeaway is that an attacker would have to do "a lot" of guesses.

piegamesde commented:
> but we don't say that "in order to roll a 3 there will be on average 5 failures"

the correct phrase here would be "for every 3 roll there will be on average 5 failures".

Anyways, I've reworked the paragraph again. I removed the 45000 number, because it doesn't help that it's correct if it's not intuitive to interpret what it represents, and the new version still gets the relevant point across IMO.
